lib/zz.mli

```diff
@@ -0,0 +1,183 @@
+(** {1 ZLIB layer.}
+
+    ZLIB is a standard on top of RFC1951. It uses [dd] implementation with LZ77
+   compression algorithm. Module provides non-blocking streaming codec to
+   {{:#decode}decode} and {{:#encode}encode} ZLIB encoding. It can efficiently
+   work payload by payload without blocking IO. *)
+
+module Bigarray = Bigarray_compat
+(** MirageOS compatibility. *)
+
+type bigstring = (char, Bigarray.int8_unsigned_elt, Bigarray.c_layout) Bigarray.Array1.t
+(** The type for [bigstring]. *)
+
+type window = Dd.window
+(** The type for windows. *)
+
+val io_buffer_size : int
+(** IO_BUFFER_SIZE 4.0.0 *)
+
+(** {2:decode ZLIB Decoder.}
+
+    Despite [dd], [zz] provides a referentially transparent {!M.decoder}. The client
+   must use [decoder] given {b by} {!M.decode} instead [decoder] given {b to}
+   {!M.decode}. A common use of [zz] is:
+
+     {[
+let rec go d0 = match M.decod d0 with
+  | `Await d1 -> ... go d1
+  | `Flush d1 -> ... go d1
+  | _ -> ... in
+     ]} *)
+
+module M : sig
+  type decoder
+  (** The type for decoders. *)
+
+  type src = [ `Channel of in_channel | `String of string | `Manual ]
+  (** The type for input sources. With a [`Manual] source the client must
+     provide input with {!src}. With [`String] or [`Channel] source the client
+     can safely discard [`Await] case (with [assert false]). *)
+
+  type decode = [ `Await of decoder | `Flush of decoder | `End of decoder | `Malformed of string ]
+
+  val decoder : src -> o:bigstring -> allocate:(int -> window) -> decoder
+  (** [decoder src ~o ~allocate] is a decoder that inputs from [src].
+
+      {b Output buffer.}
+
+      [zz], as [dd], uses [o] buffer as internal buffer to store output. We
+     recommend to allocate an {!io_buffer_size} buffer as output buffer. Then,
+     {!dst_rem} gives you how many bytes it remains in [o].
+
+      {b Window.}
+
+      ZLIB has an header to notice length of window needed to inflate given
+     input. When [zz] knows that, it calls [allocate] with a number [bits] so
+     that [1 lsl bits] is length of window. [bits] can not be upper than 15 and
+     lower than 8. [allocate] can be [fun bits -> Dd.make_window ~bits] or a
+     previously allocated window. [decoder] will take the ownership on it! *)
+
+  val decode : decoder -> decode
+  (** [decode d0] is:
+
+      {ul
+      {- [`Await d1] if [d0] has a [`Manual] input source and awaits for more
+     input. The client must use {!src} with [d1] to provide it.}
+      {- [`Flush d1] if given output buffer [o] (see {!decoder}) is full. The
+     client must use {!flush} with [d1] to {b completely} flush [o]. {!dst_rem}
+     gives you how many bytes it remains in [o]. [M.dst_rem d1 - bigstring_length o]
+     gives you how many bytes are available.}
```

Comment:
He should flushes only `bigstring_length o - M.dst_rem d1`.

Reply:
What do you think about this change, would that be correct?

```ocaml
(*
      {- [`Flush d1] if given output buffer [o] (see {!decoder}) needs to be drained before
         work can be resumed. The client must use {!flush} with [d1] to {b completely} flush [o].
         Usually [o] will be full and consist fully of bytes that need to be copied from the buffer,
         but sometimes only the first part of the buffer is used.
         In those cases  {!dst_rem} will give you the amount of free/unused bytes remain in [o].
         These should {b not} be copied since their contents are not part of the output.
         Instead the first [bigstring_length o - M.dst_rem d1] bytes should be copied when flushing [o].
         }
*)
```